swayfmt/src/comments.rs

```diff
@@ -1,26 +1,77 @@
 use std::{fmt::Write, ops::Range};
+use sway_ast::token::{Comment, CommentKind};
 use sway_types::Spanned;
 
-use crate::{formatter::FormattedCode, Formatter, FormatterError};
+use crate::{
+    formatter::FormattedCode, utils::map::comments::CommentMap, Formatter, FormatterError,
+};
+
+pub type UnformattedCode = String;
+
+#[derive(Debug, Default, Clone)]
+pub struct CommentsContext {
+    /// A BTreeMap of the mapping ByteSpan->Comment for inserting comments.
+    pub map: CommentMap,
+    /// Original unformatted code that the formatter tries to format.
+    /// The Formatter requires this to preserve newlines between comments.
+    unformatted_code: UnformattedCode,
+}
+
+impl CommentsContext {
+    pub fn new(map: CommentMap, unformatted_code: UnformattedCode) -> Self {
+        Self {
+            map,
+            unformatted_code,
+        }
+    }
+    pub fn unformatted_code(&self) -> &str {
+        &self.unformatted_code
+    }
+}
 
 pub fn has_comments<I: Iterator>(comments: I) -> bool {
     comments.peekable().peek().is_some()
 }
 
+/// This function collects newlines to insert after the comment span to preserve them.
+pub fn collect_newlines_after_comment_span(unformatted_code: &str) -> String {
```

Comment:
Perhaps we should clarify in the comment that the given `unformatted_code` should only be the unformatted code of the slice between the comment end and the next item or comment?

Reply:
Hmm actually, let me rename this function. It doesn't really make sense to be named `after_comment_span` if i'm not passing in a comment here.

Reply:
renamed + refactored it to take in a `&CommentsContext` and a `&Comment` instead of a raw `&str`